Source/JavaScriptCore/llint/InPlaceInterpreter.asm

```diff
@@ -0,0 +1,1300 @@
+# system return registers, and on the stack if not possible. After this, we perform cleanup logic to reset the stack to its
+# original state, and return to the caller.
+#
+
+#################################
+# Register and Size Definitions #
+#################################
+
+# PC = t4
+const MC = t5  # Metadata counter (index into metadata)
+const PL = t6  # Pointer to locals (index into locals)
+const PM = metadataTable
+
+if ARM64 or ARM64E
+    const IB = t7  # instruction base
+end
+
+# TODO: SIMD support, since locals will need double the space. Can we do it only sometimes?
+# May just need to write metadata that rewrites offsets. May be worth the space savings.
+# Actually, what if we just use the same thing but have a SIMD section separately allocated that
+# is "pointed" to by the 8B entries on the stack? Easier and we only need to allocate SIMD when we need
+# instead of blowing up the stack. Argument copying a little trickier though.
+
+const PtrSize = constexpr (sizeof(void*))
+const MachineRegisterSize = constexpr (sizeof(CPURegister))
+const SlotSize = constexpr (sizeof(Register))
+const LocalSize = SlotSize
+const StackValueSize = 16
+
+if X86_64 or ARM64 or ARM64E or RISCV64
+    const wasmInstance = csr0
+    const memoryBase = csr3
+    const boundsCheckingSize = csr4
+elsif ARMv7
+    const wasmInstance = csr0
+    const memoryBase = invalidGPR
+    const boundsCheckingSize = invalidGPR
+else
+end
+
+const WasmCodeBlock = CallerFrame - constexpr Wasm::numberOfIPIntCalleeSaveRegisters * SlotSize - MachineRegisterSize
+
+##########
+# Macros #
+##########
+
+# Callee Save
+
+macro saveIPIntRegisters()
+    subp 2*CalleeSaveSpaceStackAligned, sp
+    if ARM64 or ARM64E
+        storepairq wasmInstance, PB, -16[cfr]
+        storep PM, -24[cfr]
+    elsif X86_64 or RISCV64
+        storep PB, -0x8[cfr]
+        storep wasmInstance, -0x10[cfr]
+        storep PM, -0x18[cfr]
+    else
+    end
+end
+
+macro restoreIPIntRegisters()
+    if ARM64 or ARM64E
+        loadpairq -16[cfr], wasmInstance, PB
+        loadp -24[cfr], PM
+    elsif X86_64 or RISCV64
+        loadp -0x8[cfr], PB
+        loadp -0x10[cfr], wasmInstance
+        loadp -0x18[cfr], PM
+    else
+    end
+    addp 2*CalleeSaveSpaceStackAligned, sp
+end
+
+# Get IPIntCallee object at startup
+
+macro getIPIntCallee()
+    loadp Callee[cfr], ws0
+if JSVALUE64
+    andp ~(constexpr JSValue::WasmTag), ws0
+end
+    leap WTFConfig + constexpr WTF::offsetOfWTFConfigLowestAccessibleAddress, ws1
+    loadp [ws1], ws1
+    addp ws1, ws0
+    storep ws0, WasmCodeBlock[cfr]
+end
+
+# Tail-call dispatch
+
+macro advancePC(amount)
+    addq amount, PC
+end
+
+macro advancePCByReg(amount)
+    addq amount, PC
+end
+
+macro advanceMC(amount)
+    addq amount, MC
+end
+
+macro nextIPIntInstruction()
+    loadb [PB, PC, 1], t0
+if ARM64 or ARM64E
+    emit "add x0, x7, x0, lsl #8"
+    emit "br x0"
```

Comment:
We need to be very careful here. This is an unauthenticated branch. An attacker that can control metadata can now make us do arbitrary control flow.  We need to assert that no other code sits inside the 255*256 byte region around x7, and we need to make sure that x7 is always signed when it is spilled

Reply:
64KB after `x7` has been filled with either the instruction handlers or with `brk`s in most recent push. `x7` will not ever spill to stack.